syntax = "proto2";

package com.scalableminds.fossildb.proto;

message HealthRequest {}
message HealthReply {}

message GetRequest {
    required string collection = 1;
    required string key = 2;
    optional uint64 version = 3;
    optional bool mayBeEmpty = 4;
}

message GetReply {
    required bytes value = 1;
    required uint64 actualVersion = 2;
}

message PutRequest {
    required string collection = 1;
    required string key = 2;
    optional uint64 version = 3;
    required bytes value = 4;
}

message PutReply {}

message DeleteRequest {
    required string collection = 1;
    required string key = 2;
    required uint64 version = 3;
}

message DeleteReply {}

message GetMultipleVersionsRequest {
    required string collection = 1;
    required string key = 2;
    optional uint64 newestVersion = 4;
    optional uint64 oldestVersion = 3;
}

message GetMultipleVersionsReply {
    repeated bytes values = 1;
    repeated uint64 versions = 2;
}

message GetMultipleKeysRequest {
    required string collection = 1;
    required string key = 2;
    optional string prefix = 3;
    optional uint64 version = 4;
}

message GetMultipleKeysReply {
    repeated string keys = 1;
    repeated bytes values = 2;
}

message DeleteMultipleVersionsRequest {
    required string collection = 1;
    required string key = 2;
    optional uint64 newestVersion = 4;
    optional uint64 oldestVersion = 3;
}

message DeleteMultipleVersionsReply {}

message ListKeysRequest {
    required string collection = 1;
    optional uint32 limit = 2;
    optional string startAfterKey = 3;
}

message ListKeysReply {
    repeated string keys = 1;
}

message ListVersionsRequest {
    required string collection = 1;
    required string key = 2;
    optional uint32 limit = 3;
    optional uint32 offset = 4;
}

message ListVersionsReply {
    repeated uint64 versions = 1;
}


message BackupRequest {}

message BackupReply {
    required uint32 id = 1;
    required uint64 timestamp = 2;
    required uint64 size = 3;
}

message RestoreFromBackupRequest {}

message RestoreFromBackupReply {}


service FossilDB {
    rpc Health (HealthRequest) returns (stream HealthReply) {}
    rpc Get (GetRequest) returns (stream GetReply) {}
    rpc GetMultipleVersions (GetMultipleVersionsRequest) returns (stream GetMultipleVersionsReply) {}
    rpc GetMultipleKeys (GetMultipleKeysRequest) returns (stream GetMultipleKeysReply) {}
    rpc Put (PutRequest) returns (stream PutReply) {}
    rpc Delete (DeleteRequest) returns (stream DeleteReply) {}
    rpc DeleteMultipleVersions (DeleteMultipleVersionsRequest) returns (stream DeleteMultipleVersionsReply) {}
    rpc ListKeys (ListKeysRequest) returns (stream ListKeysReply) {}
    rpc ListVersions (ListVersionsRequest) returns (stream ListVersionsReply) {}
    rpc Backup (BackupRequest) returns (stream BackupReply) {}
    rpc RestoreFromBackup (RestoreFromBackupRequest) returns (stream RestoreFromBackupReply) {}
}